backend/src/knowledge_table_api/services/vector_index/base.py

```diff
@@ -0,0 +1,136 @@
+import re
+import uuid
+from abc import ABC, abstractmethod
+from typing import Any, Dict, List
+
+import numpy as np
+from langchain.schema import Document
+from pydantic import BaseModel, Field
+
+from knowledge_table_api.models.query import Rule, VectorResponse
+from knowledge_table_api.services.llm import get_keywords
+from knowledge_table_api.services.llm_service import LLMService
+
+
+class Metadata(BaseModel, extra="forbid"):
+    text: str
+    page_number: int
+    chunk_number: int
+    document_id: str
+    uuid: str = Field(default_factory=lambda: str(uuid.uuid4()))
+
+
+class VectorIndex(ABC):
+    @abstractmethod
+    async def upsert_vectors(
+        self, document_id: str, chunks: List[Document], llm_service: LLMService
+    ) -> Dict[str, str]:
+        pass
+
+    @abstractmethod
+    async def vector_search(
+        self, queries: List[str], document_id: str, llm_service: LLMService
+    ) -> dict[str, Any]:
```

Comment:
To ensure consistency, it makes sense to consider returning `VectorResponse` here. This change would provide a structured and validated response, enhancing code readability and type safety.